middle_end/flambda2/docs/inlining.md

```diff
@@ -0,0 +1,286 @@
+# Inlining in flambda2
+
+## Introduction
+
+The inliner is an important optimization that allow the user to write near 
+zero-cost abstractions easily. 
+Inlining occurs on function calls. Upon seeing a call the inliner can decide to
+*inline* it, meaning remove the call and replace it by the body of the function
+being called. Any information about the arguments of the call can then be used
+to simplify the body. Thus, the inliner can indrectly remove allocations, unused
+code and simplify some operations.
+The hard part of inlining is on deciding what is worth inlining and what isn't. 
+This document will go through the inner work of the inliner in Flambda2.
+
+Flambda2's inliner must be:
+- *Predictable*: the user should never be surprised by a decision taken by the
+  inliner, and as a corollary any decision must be explainable.
+- *Fast*: The inliner should be fairly efficient.
+
+
+## Cost_metric
+
+Two metrics are computed for all expression going through the simplifier. The 
+first one, the code size, is a rough approximation of the size of the generated
+code. The second one, the removed_operations is a set of counters tracking the
+number of operations removed during by simplifying this expression..
+
+### Code size
+
+The code size is computed in an arbitrary unit. It is not meant to be exact, 
+only to be correlated with the actual size of the generated code.
+Given an expression the code size is computed as follow:
+- If it is `Switch`, then the size is 5 times the number of arms 
+- If it is `Apply_cont`, the size is 0 if there is no trap,  5 otherwise
+- If it is `Apply` then the size depends on the type of calls. Direct calls have
+  a smaller size than indirect calls, which has a smaller size than allocating
+  extcalls.
+- `Let` has the size of its body plus the size of its expression
+- The size of a `Prim` depends on the specific primitive
+- Finally the size of a set of closures corresponds to the sum of the size of 
+  its closures, to which is added the size of an allocation and 2 if the closure
+  has arity 1, 3 otherwise. The sum of the size of it closures is involved in
+  the computation as if we were to inline a function where a set of closures was
+  defined then we will have to also copy the body of those closures.
+
+
+### Removed operations
+
+While simplifying an expression the simplifier might remove some operations. As
+all instructions are not made equal in regard to performance, it's generally a
+good idea to give a small bonus to functions that when inline removes 
+known-to-be costly expressions.
+
+The `Removed_operations` metrics are tracking the number of function calls, 
+allocations, primitives, branches, of indirect calls converted to a direct 
+calls, and of polymorphic comparisons that were removed while simplifying the
```

Comment:
Polymorphic comparisons are specialized rather than removed; well, they are specialized into non-polymorphic comparisons, so technically, the polymorphic comparison has been removed, but I fear just stating removed might be a bit confusing to people.